src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                    SkipWhiteSpace();
+                    if (_consumed >= (uint)localCopy.Length)
+                        return true;
+                }
+
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Default)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.AllowComments)
+                    {
+                        if (_tokenType == JsonTokenType.Comment || localCopy[_consumed] == JsonConstants.Solidus)
+                            return true;
+                    }
+                    else
+                    {
+                        // JsonCommentHandling.SkipComments
+                        if (localCopy[_consumed] == JsonConstants.Solidus)
+                            return true;
+                    }
+                }
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndAfterSingleJson, localCopy[_consumed]);
+            }
+            return true;
+        }
+
+        private void SkipWhiteSpace()
+        {
+            //Create local copy to avoid bounds checks.
+            ReadOnlySpan<byte> localCopy = _buffer;
+            for (; _consumed < localCopy.Length; _consumed++)
+            {
+                byte val = localCopy[_consumed];
+                if (val != JsonConstants.Space &&
+                    val != JsonConstants.CarriageReturn &&
+                    val != JsonConstants.LineFeed &&
+                    val != JsonConstants.Tab)
+                {
+                    break;
+                }
+
+                if (val == JsonConstants.LineFeed)
+                {
+                    _lineNumber++;
+                    _lineBytePosition = 0;
+                }
+                else
+                {
+                    _lineBytePosition++;
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method contains the logic for processing the next value token and determining
+        /// what type of data it is.
+        /// </summary>
+        private bool ConsumeValue(byte marker)
+        {
+            if (marker == JsonConstants.Quote)
+                return ConsumeString();
+            else if (marker == JsonConstants.OpenBrace)
+                StartObject();
+            else if (marker == JsonConstants.OpenBracket)
+                StartArray();
+            else if ((uint)(marker - '0') <= '9' - '0' || marker == '-')
+                return ConsumeNumber();
+            else if (marker == 'f')
+                return ConsumeLiteral(JsonConstants.FalseValue, JsonTokenType.False);
+            else if (marker == 't')
+                return ConsumeLiteral(JsonConstants.TrueValue, JsonTokenType.True);
+            else if (marker == 'n')
+                return ConsumeLiteral(JsonConstants.NullValue, JsonTokenType.Null);
+            else
+            {
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Default)
```

Comment:
The nesting here makes it a bit hard to see the behavior.  Consider inverting the check to just:
```C#
if (_readerOptions.CommentHandling == JsonCommentHandling.Default)
{
    Throw...
}

... // everything currently in the if block
```
or even just using a switch statement, since you have behavior for each of the three cases.